runtime.MessageSender origin parameter is inadvertently lowercased

elijahsawyers opened the pull request:
#### 643b0bd007fb4de8eb20374ed128f243ac7c1431
<pre>
runtime.MessageSender origin parameter is inadvertently lowercased
<a href="https://bugs.webkit.org/show_bug.cgi?id=288511">https://bugs.webkit.org/show_bug.cgi?id=288511</a>
<a href="https://rdar.apple.com/140291738">rdar://140291738</a>

Reviewed by Timothy Hatcher.

This patch fixes a bug where the runtime.MessageSender origin parameter would be
lowercased, which differs from the result returned from runtime.getURL.

To fix this, we can check if the origin is an extension origin, and if so,
maintain the case of the baseURL of the extension.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm:
(WebKit::WebExtensionAPIPort::sender):
Pass in the extension&apos;s baseURL.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::toWebAPI):
Modify this method to maintain the case of the extension&apos;s baseURL, if the
origin is an extension origin.

(WebKit::WebExtensionContextProxy::internalDispatchRuntimeMessageEvent):
Pass in the extension&apos;s baseURL.

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
Modify this method to take in the extension&apos;s baseURL.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIRuntime, SendMessageFromOptionsPage)):
Write a new test to validate the change.

Canonical link: <a href="https://commits.webkit.org/291118@main">https://commits.webkit.org/291118@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5e5822341eaf49e1e0dae0609e5cfd5794f7ba16

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91910 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11435 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/987 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96851 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42522 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93960 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11796 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19925 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70531 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28022 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94911 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8997 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83252 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50860 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8728 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/847 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41736 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79051 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/847 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98878 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19038 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14058 "Found 1 new test failure: ipc/large-vector-allocate-failure-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79559 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19290 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79098 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78786 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23327 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/632 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12079 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14621 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19019 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24231 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18716 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22175 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20468 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -577,8 +577,15 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
+        auto securityOrigin = (NSString *)WebCore::SecurityOrigin::create(parameters.url)->toString();
+        auto baseURLOrigin = (NSString *)makeString(baseURL.protocol(), "://"_s, baseURL.host());
+
+        if ([securityOrigin caseInsensitiveCompare:baseURLOrigin] == NSOrderedSame)
```
```suggestion
        auto securityOrigin = WebCore::SecurityOrigin::create(parameters.url)->toString();
        auto baseURLOrigin = makeString(baseURL.protocol(), "://"_s, baseURL.host());

        if (equalIgnoringASCIICase(securityOrigin, baseURLOrigin))
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
```diff
@@ -562,7 +562,7 @@ - (instancetype)initWithAggregator:(WebKit::ReplyCallbackAggregator&)aggregator
 }
 
-NSDictionary *toWebAPI(const WebExtensionMessageSenderParameters& parameters)
+NSDictionary *toWebAPI(const WebExtensionMessageSenderParameters& parameters, URL baseURL)
```
```suggestion
NSDictionary *toWebAPI(const WebExtensionMessageSenderParameters& parameters, const URL& baseURL)
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
```diff
@@ -120,7 +120,7 @@ class WebExtensionAPIWebPageRuntime : public WebExtensionAPIObject, public WebEx
 };
 
-NSDictionary *toWebAPI(const WebExtensionMessageSenderParameters&);
+NSDictionary *toWebAPI(const WebExtensionMessageSenderParameters&, URL baseURL);
```
```suggestion
NSDictionary *toWebAPI(const WebExtensionMessageSenderParameters&, const URL& baseURL);
```

elijahsawyers committed:
runtime.MessageSender origin parameter is inadvertently lowercased
https://bugs.webkit.org/show_bug.cgi?id=288511
rdar://140291738

Reviewed by Timothy Hatcher.

This patch fixes a bug where the runtime.MessageSender origin parameter would be
lowercased, which differs from the result returned from runtime.getURL.

To fix this, we can check if the origin is an extension origin, and if so,
maintain the case of the baseURL of the extension.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPortCocoa.mm:
(WebKit::WebExtensionAPIPort::sender):
Pass in the extension's baseURL.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::toWebAPI):
Modify this method to maintain the case of the extension's baseURL, if the
origin is an extension origin.

(WebKit::WebExtensionContextProxy::internalDispatchRuntimeMessageEvent):
Pass in the extension's baseURL.

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
Modify this method to take in the extension's baseURL.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
(TestWebKitAPI::TEST(WKWebExtensionAPIRuntime, SendMessageFromOptionsPage)):
Write a new test to validate the change.

Canonical link: https://commits.webkit.org/291118@main
